### ISSUE TITLE
Dropout segfaults with -OO

### ISSUE THREAD

**Issue by elbamos** [pytorch/pytorch#1848]
Example:  

```
(pytorch) amos@node:/mnt/data2/projects/nextgen/model$ python -OO ./testdropout.py
Segmentation fault (core dumped)
(pytorch) amos@node:/mnt/data2/projects/nextgen/model$ cat ./testdropout.py
from torch.autograd import Variable
import torch.nn.functional as F
import torch

ted = Variable(torch.randn(5, 4, 20))

out = F.dropout(ted, 0.2, True)

print(out)
```

Ubuntu 16.04, pytorch from last week.

**Comment by fmassa** [comment]
I managed to reproduce it with pytorch compiled from master today.
Actually, it's not limited to `dropout`, but other functions give the same behavior as well (`relu`, `sigmoid`), and the `-O` flag is enough to make it segfault.

**Comment by elbamos** [comment]
Pls let me know if there's a way I can be helpful.

**Comment by soumith** [comment]
@fmassa @elbamos did anyone investigate this further? if you run it in gdb, where is the segfault showing?

**Comment by elbamos** [comment]
@soumith I no longer have access to the system that was generating the error (changed jobs).  I can try to reproduce on OS X, but OS X currently can't install from Master.

**Comment by fmassa** [comment]
@soumith I don't have the computer with which I reproduced it with me this weekend. If I recall properly, there was no stack trace from gdb, but I'm not 100% sure anymore.

**Comment by howard0su** [comment]
I happened run to this as well. I use the same python script to repro under gdb. here is output:
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
PyObject_IsInstance () at ../Objects/abstract.c:2947
2947	../Objects/abstract.c: No such file or directory.
(gdb) where
#0  PyObject_IsInstance () at ../Objects/abstract.c:2947
#1  0x00007fffcbaedb9c in THPFunction_new (type=<optimized out>, args=<optimized out>, kwargs=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:299
#2  0x00000000004b6693 in type_call.lto_priv () at ../Objects/typeobject.c:749
#3  0x00000000004b0cb3 in PyObject_Call () at ../Objects/abstract.c:2546
#4  0x00000000004c9faf in do_call (nk=<optimized out>, na=3, pp_stack=0x7fffffffdf00, func=0xae6df0) at ../Python/ceval.c:4567
#5  call_function (oparg=<optimized out>, pp_stack=0x7fffffffdf00) at ../Python/ceval.c:4372
#6  PyEval_EvalFrameEx () at ../Python/ceval.c:2987
#7  0x00000000004c2765 in PyEval_EvalCodeEx () at ../Python/ceval.c:3582
#8  0x00000000004ca099 in fast_function (nk=0, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffe110, func=0x7fff85093b18)
    at ../Python/ceval.c:4445
#9  call_function (oparg=<optimized out>, pp_stack=0x7fffffffe110) at ../Python/ceval.c:4370
#10 PyEval_EvalFrameEx () at ../Python/ceval.c:2987
#11 0x00000000004c2765 in PyEval_EvalCodeEx () at ../Python/ceval.c:3582
#12 0x00000000004c2509 in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:669
#13 0x00000000004f1def in run_mod.lto_priv () at ../Python/pythonrun.c:1376
#14 0x00000000004ec652 in PyRun_FileExFlags () at ../Python/pythonrun.c:1362
#15 0x00000000004eae31 in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:948
#16 0x000000000049e14a in Py_Main () at ../Modules/main.c:640
#17 0x00007ffff7810830 in __libc_start_main (main=0x49dab0 <main>, argc=3, argv=0x7fffffffe558, init=<optimized out>, fini=<optimized out>,
    rtld_fini=<optimized out>, stack_end=0x7fffffffe548) at ../csu/libc-start.c:291
#18 0x000000000049d9d9 in _start ()
(gdb) up
#1  0x00007fffcbaedb9c in THPFunction_new (type=<optimized out>, args=<optimized out>, kwargs=<optimized out>)
    at torch/csrc/autograd/python_function.cpp:299
299	  self->cdata.is_stochastic = PyObject_IsInstance(obj, THPStochasticFunctionClass);
(gdb) info locals
obj = 0x7fffcdc4caa0
self = 0x7fffcdc4caa0
(gdb) p obj
$1 = (PyObject *) 0x7fffcdc4caa0
(gdb) p *obj
$2 = {
  ob_refcnt = 1,
  ob_type = 0xae6df0
}
(gdb) p *self
$3 = {
  ob_refcnt = 1,
  ob_type = 0xae6df0,
  needs_input_grad = 0x0,
  to_save = 0x0,
  shared_pairs = 0x0,
  non_differentiable = 0x0,
  dirty_tensors = 0x0,
  output_info = 0x0,
  saved_variables = 0x0,
  is_variable_input = 0x0,
  has_freed_buffers = 0 '\000',
  cdata = {
    <torch::autograd::Function> = {
      _vptr.Function = 0x7fffcc90c780 <vtable for torch::autograd::PyFunction+16>,
      num_inputs = -1,
      next_functions = std::vector of length 0, capacity 0,
      is_executable = false,
      is_stochastic = false,
      pre_hooks = std::vector of length 0, capacity 0,
      post_hooks = std::vector of length 0, capacity 0
    },
    members of torch::autograd::PyFunction:
    obj = 0x7fffcdc4caa0
  }
}

**Comment by soumith** [comment]
fixed via #2120
